
<table>

<tr valign="top">
<td align="right" class="bibtexnumber">
[<a name="agresti_approximate_1998">1</a>]
</td>
<td class="bibtexitem">
Alan Agresti and Brent&nbsp;A. Coull.
 Approximate is better than “exact” for interval estimation of
  binomial proportions.
 <em>The American Statistician</em>, 52(2):119–126, 1998.
[&nbsp;<a href="http://www.tandfonline.com/doi/abs/10.1080/00031305.1998.10480550">http</a>&nbsp;]

</td>
</tr>


<tr valign="top">
<td align="right" class="bibtexnumber">
[<a name="agresti_simple_2000">2</a>]
</td>
<td class="bibtexitem">
Alan Agresti and Brian Caffo.
 Simple and effective confidence intervals for proportions and
  differences of proportions result from adding two successes and two failures.
 <em>The American Statistician</em>, 54(4):280–288, 2000.
[&nbsp;<a href="http://amstat.tandfonline.com/doi/full/10.1080/00031305.2000.10474560">http</a>&nbsp;]

</td>
</tr>


<tr valign="top">
<td align="right" class="bibtexnumber">
[<a name="piegorsch_sample_2004">3</a>]
</td>
<td class="bibtexitem">
Walter&nbsp;W. Piegorsch.
 Sample sizes for improved binomial confidence intervals.
 <em>Computational Statistics &amp; Data Analysis</em>, 46(2):309-316, June
  2004.
[&nbsp;<a href="http://dx.doi.org/10.1016/j.csda.2003.10.002">DOI</a>&nbsp;| 
<a href="http://www.sciencedirect.com/science/article/pii/S0167947303002354">http</a>&nbsp;]

</td>
</tr>


<tr valign="top">
<td align="right" class="bibtexnumber">
[<a name="benjamini_false_2005">4</a>]
</td>
<td class="bibtexitem">
Yoav Benjamini and Daniel Yekutieli.
 False discovery rate–adjusted multiple confidence intervals for
  selected parameters.
 <em>Journal of the American Statistical Association</em>,
  100(469):71–81, 2005.
[&nbsp;<a href="http://amstat.tandfonline.com/doi/abs/10.1198/016214504000001907">http</a>&nbsp;]

</td>
</tr>


<tr valign="top">
<td align="right" class="bibtexnumber">
[<a name="schaarschmidt_approximate_2008">5</a>]
</td>
<td class="bibtexitem">
Frank Schaarschmidt, Martin Sill, and Ludwig&nbsp;A. Hothorn.
 Approximate simultaneous confidence intervals for multiple contrasts
  of binomial proportions.
 <em>Biometrical Journal</em>, 50(5):782–792, 2008.
[&nbsp;<a href="http://dx.doi.org/10.1002/bimj.200710465">DOI</a>&nbsp;| 
<a href="http://onlinelibrary.wiley.com/doi/10.1002/bimj.200710465/abstract">http</a>&nbsp;]

</td>
</tr>


<tr valign="top">
<td align="right" class="bibtexnumber">
[<a name="fagerland_recommended_2011">6</a>]
</td>
<td class="bibtexitem">
Morten&nbsp;W. Fagerland, Stian Lydersen, and Petter Laake.
 Recommended confidence intervals for two independent binomial
  proportions.
 <em>Statistical methods in medical research</em>, 2011.
[&nbsp;<a href="http://smm.sagepub.com/content/early/2011/10/11/0962280211415469.abstract">http</a>&nbsp;]

</td>
</tr>


<tr valign="top">
<td align="right" class="bibtexnumber">
[<a name="knol_misuse_2011">7</a>]
</td>
<td class="bibtexitem">
Mirjam&nbsp;J. Knol, Wiebe&nbsp;R. Pestman, and Diederick&nbsp;E. Grobbee.
 The (mis)use of overlap of confidence intervals to assess effect
  modification.
 <em>European Journal of Epidemiology</em>, 26(4):253-254, April 2011.
 PMID: 21424218 PMCID: PMC3088813.
[&nbsp;<a href="http://dx.doi.org/10.1007/s10654-011-9563-8">DOI</a>&nbsp;| 
<a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3088813/">http</a>&nbsp;]

</td>
</tr>


<tr valign="top">
<td align="right" class="bibtexnumber">
[<a name="gerstung_reliable_2012">8</a>]
</td>
<td class="bibtexitem">
Moritz Gerstung, Christian Beisel, Markus Rechsteiner, Peter Wild, Peter
  Schraml, Holger Moch, and Niko Beerenwinkel.
 Reliable detection of subclonal single-nucleotide variants in tumour
  cell populations.
 <em>Nature Communications</em>, 3:811, May 2012.
[&nbsp;<a href="http://dx.doi.org/10.1038/ncomms1814">DOI</a>&nbsp;| 
<a href="http://www.nature.com/ncomms/journal/v3/n5/full/ncomms1814.html">.html</a>&nbsp;]

</td>
</tr>


<tr valign="top">
<td align="right" class="bibtexnumber">
[<a name="yates_evolution_2012">9</a>]
</td>
<td class="bibtexitem">
Lucy&nbsp;R. Yates and Peter&nbsp;J. Campbell.
 Evolution of the cancer genome.
 <em>Nature Reviews Genetics</em>, 13(11):795, November 2012.
[&nbsp;<a href="http://dx.doi.org/10.1038/nrg3317">DOI</a>&nbsp;| 
<a href="http://www.nature.com/nrg/journal/v13/n11/abs/nrg3317.html">.html</a>&nbsp;]

</td>
</tr>


<tr valign="top">
<td align="right" class="bibtexnumber">
[<a name="fleiss_statistical_2013">10</a>]
</td>
<td class="bibtexitem">
Joseph&nbsp;L. Fleiss, Bruce Levin, and Myunghee&nbsp;Cho Paik.
 <em>Statistical methods for rates and proportions</em>.
 John Wiley &amp; Sons, 2013.
[&nbsp;<a href="http://books.google.de/books?hl=de&lr=&id=9VefO7a8GeAC&oi=fnd&pg=PT13&dq=Interval+Estimation+for+the+Difference+Between+Independent+Proportions&ots=po7iNMzBKT&sig=eSFnYXR3AbF3F4BbxIWpGfDUFHM">http</a>&nbsp;]

</td>
</tr>


<tr valign="top">
<td align="right" class="bibtexnumber">
[<a name="pyl_h5vcdata:_2013">11</a>]
</td>
<td class="bibtexitem">
Paul&nbsp;Theodor Pyl.
 h5vcData: example data for the h5vc package, 2013.
[&nbsp;<a href="http://bioconductor.org/packages/release/data/experiment/html/h5vcData.html">.html</a>&nbsp;]

</td>
</tr>


<tr valign="top">
<td align="right" class="bibtexnumber">
[<a name="decrouez_split_2013">12</a>]
</td>
<td class="bibtexitem">
Geoffrey Decrouez and Peter Hall.
 Split sample methods for constructing confidence intervals for
  binomial and poisson parameters.
 <em>Journal of the Royal Statistical Society: Series B (Statistical
  Methodology)</em>, page n/a–n/a, 2013.
[&nbsp;<a href="http://dx.doi.org/10.1111/rssb.12051">DOI</a>&nbsp;| 
<a href="http://onlinelibrary.wiley.com/doi/10.1111/rssb.12051/abstract">http</a>&nbsp;]

</td>
</tr>


<tr valign="top">
<td align="right" class="bibtexnumber">
[<a name="agresti_categorical_2013">13</a>]
</td>
<td class="bibtexitem">
Alan Agresti.
 <em>Categorical data analysis</em>.
 Wiley, Hoboken, NJ, 2013.

</td>
</tr>


<tr valign="top">
<td align="right" class="bibtexnumber">
[<a name="kim_comparing_2013">14</a>]
</td>
<td class="bibtexitem">
Su&nbsp;Y. Kim and Terence&nbsp;P. Speed.
 Comparing somatic mutation-callers: beyond venn diagrams.
 <em>BMC Bioinformatics</em>, 14(1):189, June 2013.
 PMID: 23758877.
[&nbsp;<a href="http://dx.doi.org/10.1186/1471-2105-14-189">DOI</a>&nbsp;| 
<a href="http://www.biomedcentral.com/1471-2105/14/189/abstract">http</a>&nbsp;]

</td>
</tr>


<tr valign="top">
<td align="right" class="bibtexnumber">
[<a name="roberts_comparative_2013">15</a>]
</td>
<td class="bibtexitem">
Nicola&nbsp;D. Roberts, R.&nbsp;Daniel Kortschak, Wendy&nbsp;T. Parker, Andreas&nbsp;W. Schreiber,
  Susan Branford, Hamish&nbsp;S. Scott, Garique Glonek, and David&nbsp;L. Adelson.
 A comparative analysis of algorithms for somatic SNV detection in
  cancer.
 <em>Bioinformatics</em>, 29(18):2223-2230, September 2013.
 PMID: 23842810.
[&nbsp;<a href="http://dx.doi.org/10.1093/bioinformatics/btt375">DOI</a>&nbsp;| 
<a href="http://bioinformatics.oxfordjournals.org/content/29/18/2223">http</a>&nbsp;]

</td>
</tr>


<tr valign="top">
<td align="right" class="bibtexnumber">
[<a name="pyl_h5vc:_2014">16</a>]
</td>
<td class="bibtexitem">
Paul&nbsp;Theodor Pyl, Julian Gehring, Bernd Fischer, and Wolfgang Huber.
 h5vc: Scalable nucleotide tallies with HDF5.
 <em>Bioinformatics</em>, page btu026, January 2014.
 PMID: 24451629.
[&nbsp;<a href="http://dx.doi.org/10.1093/bioinformatics/btu026">DOI</a>&nbsp;| 
<a href="http://bioinformatics.oxfordjournals.org/content/early/2014/01/21/bioinformatics.btu026">http</a>&nbsp;]

</td>
</tr>
</table>